src/dvc_data/fs.py

```diff
@@ -85,25 +92,28 @@ def _cache_remote_file(
         cache_storage: "ObjectStorage",
         fs: "FileSystem",
         path: "AnyFSPath",
+        hash_info: Optional["HashInfo"],
     ) -> Tuple["FileSystem", "AnyFSPath"]:
         from dvc_objects.fs.local import LocalFileSystem
 
-        if isinstance(fs, LocalFileSystem):
-            return fs, path
+        odb: "HashFileDB" = cache_storage.odb
+        oid = hash_info.value if hash_info else None
+        hash_name = hash_info.name if hash_info else None
+        assert odb.hash_name
 
-        from dvc_data.hashfile.build import _upload_file
+        if isinstance(fs, LocalFileSystem) or not oid or odb.hash_name != hash_name:
+            return fs, path
 
-        cache_odb = cache_storage.odb
-        _, obj = _upload_file(path, fs, cache_odb, cache_odb)
-        return cache_odb.fs, obj.path
+        odb.add(path, fs, oid)
+        return odb.fs, odb.oid_to_path(oid)
```

Comment:
I am using `odb.add()` now which should be less costly. The `oid` is reused from metadata and is not verified. It does concern me a bit, but since index is used in other places, and push/pull never verify, it wasn't worth maintaining the previous behaviour.